### PR TITLE
tests: fix api version checks

### PR DESCRIPTION
**Pull request by airlied**
Make sure the physical device as well as the instance support
the requested version

Closes: #2666

### CODE REVIEW

**Comment by airlied** [comment]
I noticed that is was just incorrect checks when I read more code. This version fixes two tests to do correct version checks.

**Comment by airlied** [comment]
marking WIP until I get a full run completed.

File: tests/vklayertests_gpu.cpp [hunk 1]
```diff
@@ -42,6 +42,11 @@ TEST_F(VkGpuAssistedLayerTest, GpuValidationArrayOOBGraphicsShaders) {
         "GPU validation: Verify detection of out-of-bounds descriptor array indexing and use of uninitialized descriptors.");
 
     SetTargetApiVersion(VK_API_VERSION_1_1);
+    if (DeviceValidationVersion() < VK_API_VERSION_1_1) {
```
Comment (ncesario-lunarg) [inline tests/vklayertests_gpu.cpp:45]: I don't think this call makes sense before `InitGpuAssistedFramework` (i.e., I don't think `GetPhysicalDeviceProperties` has been called yet).

File: tests/vklayertests_others.cpp [hunk 2]
```diff
@@ -10866,13 +10871,14 @@ TEST_F(VkLayerTest, ValidateImportMemoryHandleType) {
 TEST_F(VkLayerTest, ValidateExtendedDynamicStateDisabled) {
     TEST_DESCRIPTION("Validate VK_EXT_extended_dynamic_state VUs");
 
-    uint32_t version = SetTargetApiVersion(VK_API_VERSION_1_1);
-    if (version < VK_API_VERSION_1_1) {
+    ASSERT_NO_FATAL_FAILURE(InitFramework(m_errorMonitor));
+
+    SetTargetApiVersion(VK_API_VERSION_1_1);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_others.cpp:10876]: Set before `InitFramework`?

File: tests/vklayertests_others.cpp [hunk 3]
```diff
@@ -10992,13 +10998,14 @@ TEST_F(VkLayerTest, ValidateExtendedDynamicStateDisabled) {
 TEST_F(VkLayerTest, ValidateExtendedDynamicStateEnabled) {
     TEST_DESCRIPTION("Validate VK_EXT_extended_dynamic_state VUs");
 
-    uint32_t version = SetTargetApiVersion(VK_API_VERSION_1_1);
-    if (version < VK_API_VERSION_1_1) {
+    ASSERT_NO_FATAL_FAILURE(InitFramework(m_errorMonitor));
+
+    SetTargetApiVersion(VK_API_VERSION_1_1);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_others.cpp:11003]: Set before `InitFramework`?

File: tests/vklayertests_others.cpp [hunk 4]
```diff
@@ -11283,13 +11290,14 @@ TEST_F(VkLayerTest, ValidateExtendedDynamicStateEnabled) {
 TEST_F(VkLayerTest, ValidateExtendedDynamicStateEnabledNoMultiview) {
     TEST_DESCRIPTION("Validate VK_EXT_extended_dynamic_state VUs");
 
-    uint32_t version = SetTargetApiVersion(VK_API_VERSION_1_1);
-    if (version < VK_API_VERSION_1_1) {
+    ASSERT_NO_FATAL_FAILURE(InitFramework(m_errorMonitor));
+
+    SetTargetApiVersion(VK_API_VERSION_1_1);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_others.cpp:11295]: Set before `InitFramework`?

File: tests/vkpositivelayertests.cpp [hunk 5]
```diff
@@ -135,13 +135,14 @@ TEST_F(VkPositiveLayerTest, NullFunctionPointer) {
 TEST_F(VkPositiveLayerTest, ViewportWithCountNoMultiViewport) {
     TEST_DESCRIPTION("DynamicViewportWithCount/ScissorWithCount without multiViewport feature not enabled.");
 
-    uint32_t version = SetTargetApiVersion(VK_API_VERSION_1_1);
-    if (version < VK_API_VERSION_1_1) {
+    ASSERT_NO_FATAL_FAILURE(InitFramework(m_errorMonitor));
+
+    SetTargetApiVersion(VK_API_VERSION_1_1);
```
Comment (ncesario-lunarg) [inline tests/vkpositivelayertests.cpp:140]: Set before `InitFramework`?

File: tests/vklayertests_best_practices.cpp [hunk 6]
```diff
@@ -720,13 +721,18 @@ TEST_F(VkBestPracticesLayerTest, TripleBufferingTest) {
...
 
-    // Vulkan 1.1 required to test vkGetPhysicalDeviceQueueFamilyProperties2
-    app_info_.apiVersion = VK_API_VERSION_1_1;
     // VK_KHR_get_physical_device_properties2 required to test vkGetPhysicalDeviceQueueFamilyProperties2KHR
     instance_extensions_.emplace_back(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
     ASSERT_NO_FATAL_FAILURE(InitBestPracticesFramework());
     const vk_testing::PhysicalDevice phys_device_obj(gpu_);
 
+    // Vulkan 1.1 required to test vkGetPhysicalDeviceQueueFamilyProperties2
+    SetTargetApiVersion(VK_API_VERSION_1_1);
```
[the remark this answers is not in the document]
Comment (airlied) [reply tests/vklayertests_best_practices.cpp]: Yes I got confused by the ordering in a few places, I think I've sorted them all out now. As the version validation check needs the gpu_ object which means it has to be after initing, but the set target needs to be before it.

File: tests/vklayertests_gpu.cpp (hunk 1 above)
Comment (airlied) [reply tests/vklayertests_gpu.cpp:48]: Fixed that now I think.

File: tests/vklayertests_best_practices.cpp [hunk 7]
```diff
@@ -140,8 +141,8 @@ TEST_F(VkBestPracticesLayerTest, UseDeprecatedInstanceExtensions) {
 TEST_F(VkBestPracticesLayerTest, UseDeprecatedDeviceExtensions) {
     TEST_DESCRIPTION("Create a device with a deprecated extension.");
 
-    uint32_t version = SetTargetApiVersion(VK_API_VERSION_1_2);
-    if (version < VK_API_VERSION_1_2) {
+    SetTargetApiVersion(VK_API_VERSION_1_2);
+    if (DeviceValidationVersion() < VK_API_VERSION_1_2) {
```
Comment (sfricke-samsung) [inline tests/vklayertests_best_practices.cpp:145]: Does this work? This relies on `VkRenderFramework`'s `physDevProps_` to get the device version which doesn't get set until `InitFramework()` is called (which `InitBestPracticesFramework()` calls as well)

Comment (airlied) [reply tests/vklayertests_best_practices.cpp:145]: the latest version I pushed should get this right now.

Comment (ncesario-lunarg) [reply tests/vklayertests_best_practices.cpp:145]: Looks like this one still needs to be moved.

File: tests/vkpositivelayertests.cpp (hunk 5 above)
Comment (ncesario-lunarg) [reply tests/vkpositivelayertests.cpp:140]: I think this still needs to move before InitFramework?

**Comment by ncesario-lunarg** [comment]
Running through [internal CI](http://erusea:4040/8768).

**Comment by ncesario-lunarg** [comment]
@airlied looks like this is failing on our AMD machine:
```
[ RUN      ] VkLayerTest.CopyInvalidImageMemory
Unexpected: Validation Error: [ VUID-vkCmdCopyImage-srcImage-01546 ] Object 0: handle = 0xf8d8ce0000000003, type = VK_OBJECT_TYPE_UNKNOWN; | MessageID = 0xa74df576 | vkCmdCopyImage(): VkImage 0xf8d8ce0000000003[] used with no memory bound. Memory should be bound by calling vkBindImageMemory(). The Vulkan spec states: If srcImage is non-sparse then the image or disjoint plane to be copied must be bound completely and contiguously to a single VkDeviceMemory object (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-srcImage-01546)
Other error messages logged for this test were:
     Validation Error: [ VUID-vkCmdCopyImage-srcImage-01546 ] Object 0: handle = 0xf8d8ce0000000003, type = VK_OBJECT_TYPE_UNKNOWN; | MessageID = 0xa74df576 | vkCmdCopyImage(): VkImage 0xf8d8ce0000000003[] used with no memory bound. Memory should be bound by calling vkBindImageMemory(). The Vulkan spec states: If srcImage is non-sparse then the image or disjoint plane to be copied must be bound completely and contiguously to a single VkDeviceMemory object (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-srcImage-01546)
C:\j\w8\32-bit\Release\tests\vkrenderframework.cpp(185): error: Failed
Did not receive expected error 'VUID-vkCmdCopyImage-srcImage-00127'
Unexpected: Validation Error: [ VUID-vkCmdCopyImage-dstImage-01547 ] Object 0: handle = 0xf8d8ce0000000003, type = VK_OBJECT_TYPE_UNKNOWN; | MessageID = 0x637cc6c6 | vkCmdCopyImage(): VkImage 0xf8d8ce0000000003[] used with no memory bound. Memory should be bound by calling vkBindImageMemory(). The Vulkan spec states: If dstImage is non-sparse then the image or disjoint plane that is the destination of the copy must be bound completely and contiguously to a single VkDeviceMemory object (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-dstImage-01547)
Unexpected: Validation Error: [ VUID-vkCmdCopyImage-srcImageLayout-00128 ] Object 0: handle = 0xa31c800, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xef9e2bd6 | vkCmdCopyImage(): Cannot use VkNonDispatchableHandle 0xf018750000000004[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_UNDEFINED that doesn't match the previously used layout VK_IMAGE_LAYOUT_GENERAL. The Vulkan spec states: srcImageLayout must specify the layout of the image subresources of srcImage specified in pRegions at the time this command is executed on a VkDevice (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-srcImageLayout-00128)
Unexpected: Validation Error: [ VUID-vkCmdCopyImage-dstImageLayout-00133 ] Object 0: handle = 0xa31c800, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xbba5299 | vkCmdCopyImage(): Cannot use VkNonDispatchableHandle 0xf8d8ce0000000003[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_GENERAL that doesn't match the previously used layout VK_IMAGE_LAYOUT_UNDEFINED. The Vulkan spec states: dstImageLayout must specify the layout of the image subresources of dstImage specified in pRegions at the time this command is executed on a VkDevice (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-dstImageLayout-00133)
Other error messages logged for this test were:
     Validation Error: [ VUID-vkCmdCopyImage-dstImage-01547 ] Object 0: handle = 0xf8d8ce0000000003, type = VK_OBJECT_TYPE_UNKNOWN; | MessageID = 0x637cc6c6 | vkCmdCopyImage(): VkImage 0xf8d8ce0000000003[] used with no memory bound. Memory should be bound by calling vkBindImageMemory(). The Vulkan spec states: If dstImage is non-sparse then the image or disjoint plane that is the destination of the copy must be bound completely and contiguously to a single VkDeviceMemory object (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-dstImage-01547)
     Validation Error: [ VUID-vkCmdCopyImage-srcImageLayout-00128 ] Object 0: handle = 0xa31c800, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xef9e2bd6 | vkCmdCopyImage(): Cannot use VkNonDispatchableHandle 0xf018750000000004[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_UNDEFINED that doesn't match the previously used layout VK_IMAGE_LAYOUT_GENERAL. The Vulkan spec states: srcImageLayout must specify the layout of the image subresources of srcImage specified in pRegions at the time this command is executed on a VkDevice (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-srcImageLayout-00128)
     Validation Error: [ VUID-vkCmdCopyImage-dstImageLayout-00133 ] Object 0: handle = 0xa31c800, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0xbba5299 | vkCmdCopyImage(): Cannot use VkNonDispatchableHandle 0xf8d8ce0000000003[] (layer=0 mip=0) with specific layout VK_IMAGE_LAYOUT_GENERAL that doesn't match the previously used layout VK_IMAGE_LAYOUT_UNDEFINED. The Vulkan spec states: dstImageLayout must specify the layout of the image subresources of dstImage specified in pRegions at the time this command is executed on a VkDevice (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdCopyImage-dstImageLayout-00133)
C:\j\w8\32-bit\Release\tests\vkrenderframework.cpp(185): error: Failed
```

File: tests/vklayertests_buffer_image_memory_sampler.cpp [hunk 8]
```diff
@@ -7574,12 +7579,22 @@ TEST_F(VkLayerTest, CopyInvalidImageMemory) {
     copy_region.extent.depth = 1;
 
     m_commandBuffer->begin();
-    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-vkCmdCopyImage-srcImage-01546");
+
+    if (sampler_ycbcr_conversion) {
+        m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-vkCmdCopyImage-srcImage-01546");
+    } else {
+        m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-vkCmdCopyImage-srcImage-00127");
+    }
```
Comment (ncesario-lunarg) [inline tests/vklayertests_buffer_image_memory_sampler.cpp:7587]: I think this (and similar changes in this file) need to be  reverted.

Stepping through this code locally on my laptop, I can see that everything works as expected. However, launching the test remotely (I tried RDP, ssh, and whatever method jenkins uses to launch processes), the same test on the same machine fails.

Comment (airlied) [reply tests/vklayertests_buffer_image_memory_sampler.cpp:7587]: I've noticed some tests fail if you don't have a DISPLAY for them to connect to.

Comment (ncesario-lunarg) [reply tests/vklayertests_buffer_image_memory_sampler.cpp:7587]: @airlied, I needed to make these changes to get things working on our AMD machine (which just makes sure `VK_KHR_sampler_ycbcr_conversion` is enabled if the device version is < 1.1): https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2877/files#diff-f6523f16c921e05a08a2224f719fdfb93d81c20d5bccebdc2b54692065a21ad3R7529-R7536. If you'd like I can just merge #2877.

**Comment by spencer-lunarg** [comment]
There was a lot of work to fix this issue, I went through all of these and they are now fixed in the top of tree so closing. Thanks again @airlied for raising this issue as it was long overdue

**Pull request closed without merging**